### PR TITLE
Add github actions for unit tests

### PR DESCRIPTION
**Pull request by hayarobi**
- Any push of any branches will trigger short unit tests.
- User can trigger normal unit tests on demand.
- Also some long running unit tests will be skipped in short unit tests

### CODE REVIEW

**Comment by kroggen** [comment]
When this PR was created, it run 2 different checks. This is not needed. It should run only the main one, made for PRs

File: .github/workflows/default_unittest.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# This workflow will build a golang project
+# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-go
+
+name:  Unit Tests
+
+run-name: Unit tests of ${{ github.ref_type }} "${{ github.ref_name }}" by @${{ github.actor }}
```
Comment (kroggen) [inline .github/workflows/default_unittest.yml:6]: This information is redundant. The list of actions already has information about user and branch

Check [here](https://github.com/aergoio/aergo/actions)

File: .github/workflows/short_unittest.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+# This workflow will build a golang project
+# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-go
+
+name:  Short Unit Tests
+
+run-name: Short unit tests of ${{ github.ref_type }} "${{ github.ref_name }}" by @${{ github.actor }}
```
Comment (kroggen) [inline .github/workflows/short_unittest.yml:6]: same

File: .github/workflows/short_unittest.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+on:
+  push:
+    branches-ignore: # master and develop run full test by other workflow
+      - master
+      - develop
```
Comment (kroggen) [inline .github/workflows/short_unittest.yml:12]: something must be added to prevent the run when a new pull request is created

File: .github/workflows/short_unittest.yml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        go-version: 1.19
+
+    - name: Install dependencies
+      run: |
+        sudo apt-get install -y jq m4 make cmake
+
+    - name: Build
+      run: make
+
+    - name: Unit Tests short
```
Comment (kroggen) [inline .github/workflows/short_unittest.yml:37]: Unit Tests (short)
or
Short Unit Tests

File: syncer/syncerservice_test.go [hunk 5]
```diff
@@ -10,7 +11,14 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+func TestMain(m *testing.M) {
+	os.Exit(m.Run())
+}
```
Comment (kroggen) [inline syncer/syncerservice_test.go:16]: What does this code do?

**Review by kroggen — CHANGES_REQUESTED** [review]

File: syncer/syncerservice_test.go (hunk 5 above)
Comment (hayarobi) [reply syncer/syncerservice_test.go:16]: It's a sign of incomplete cleanup. I initially tried to change the log level because the test was taking a lot of meaningless logs, but I gave up because it couldn't be done easily. I reverted that code and missed it. 

I'll fix those you said next week.

File: .github/workflows/build_and_test.yml [hunk 6]
```diff
@@ -1,7 +1,9 @@
 # This workflow will build a golang project
 # For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-go
 
-name: Build and Test
+name: Integration Test
+
+run-name: Integration tests and unit tests of ${{ github.ref_type }} "${{ github.ref_name }}"
```
[the remark this answers is not in the document]
Comment (hayarobi) [reply .github/workflows/build_and_test.yml:6]: I've struggled with naming these two unit test workflows. 
Do you have any suggestions of good names that would easily distinguish between the following three?
1. short unit tests 
2. all unit tests
3. all unit tests and integration tests

[the remark this answers is not in the document]
Comment (kroggen) [reply .github/workflows/build_and_test.yml:6]: I mean the `run-name` can be removed

The `name` can be kept "Build and Test" or "Complete Tests"

"Integration Tests" is when we use `aergosvr` and `aergocli` together to check how they work integrated

We can also have integration tests with herajs, herapy, heraj, libaergo... together with `aergosvr` in test mode

And then when we have many `aergosvr` running on separate nodes, and test the deploy with the IDE and calls with wallets, then it is "Functional Test"

File: .github/workflows/default_unittest.yml [hunk 7]
```diff
@@ -0,0 +1,33 @@
+# This workflow will build a golang project
+# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-go
+
+name:  Unit Tests
+
+on:
+  workflow_dispatch:
```
Comment (kroggen) [inline .github/workflows/default_unittest.yml:7]: is this file really needed? the complete test also can be triggered manually

File: .github/workflows/short_unittest.yml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+on:
+  push:
+    branches-ignore: # master and develop run full test by other workflow
+      - master
+      - develop
+
+jobs:
+  build_and_unittest:
+    # this prevents the action to run on pull requests
+    if: github.event_name == 'push' && github.ref != 'refs/pull/*/merge'
```
Comment (kroggen) [inline .github/workflows/short_unittest.yml:15]: This is not working

Comment (kroggen) [reply .github/workflows/short_unittest.yml:15]: This is an option: https://github.com/marketplace/actions/skip-duplicate-actions

**Comment by kroggen** [comment]
Another option to avoid the duplicate run of actions is to unify the 2 actions into a single one.

First by removing the filter of branches on push on `build_and_test.yaml`

Then run the steps conditionally:

```yaml
    - name: Complete Unit Test
      if: github.ref == 'refs/heads/master' || github.ref == 'refs/heads/develop'
      run: ...

    - name: Short Unit Test
      if: github.ref != 'refs/heads/master' && github.ref != 'refs/heads/develop'
      run: ...
```

But I suspect that it will still run the action again when the PR is merged...

If this is the case, then [this](https://github.com/marketplace/actions/skip-duplicate-actions) would be better

File: .github/workflows/default_unittest.yml (hunk 7 above)
Comment (hayarobi) [reply .github/workflows/default_unittest.yml:7]: I thought it depends on the size of integration tests. If running scripts `tests/run_tests.sh` takes much longer than whole unit tests. But after hearing what you said, I think it's better run it directly on the local machine when I want to test.

File: .github/workflows/short_unittest.yml (hunk 8 above)
Comment (kroggen) [reply .github/workflows/short_unittest.yml]: The reason why the 2nd action did not run (was skipped) is because it had run before, not because the first action has run.

We can see this [here](https://github.com/aergoio/aergo/actions/runs/5077771284/jobs/9121449880?pr=207) where it states:

```
Skip execution because the exact same files have been successfully checked in run https://github.com/aergoio/aergo/actions/runs/5077707153
```

But that run is from the same action

Comment (kroggen) [reply .github/workflows/short_unittest.yml]: The other option bellow (joining them into a single job) may be better, and simpler

Comment (kroggen) [reply .github/workflows/short_unittest.yml]: I will try to implement and check the result

**Comment by kroggen** [comment]
I made a test on another fork and it did not work

You can see the PR [here](https://github.com/kroggen/aergo/pull/2) it run the short unit tests instead of the complete tests on the pull request

**Review by kroggen — APPROVED** [review]
OK, it can be merged as is

**Comment by kroggen** [comment]
If you will no more use the branch, then it can be deleted
